1-raw-data/mailinglists/getMailData.go

```diff
@@ -0,0 +1,177 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+//TODO
+// Create tests
+// check the most recent file stored and pull only what isn't there
+// Run this monthly at start of new month to pull all new data
+
+import (
+	"cloud.google.com/go/storage"
+	"context"
+	"flag"
+	"fmt"
+	"github.com/PuerkitoBio/goquery"
+	"google.golang.org/api/iterator"
+	"io"
+	"log"
+	"net/http"
+	"strings"
+)
+
+type GCS struct {
+	bucketName string
+	ctx context.Context
+	client *storage.Client
+	bucket *storage.BucketHandle
+}
+
+func connectCTX() context.Context{
+	ctx := context.Background()
+	//ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+	//defer cancel()
+	return ctx
+}
+
+func (gcs *GCS) connectGCS() {
+	client, err := storage.NewClient(gcs.ctx)
+	if err != nil {
+		log.Fatalf("Failed to create client: %v", err)
+	}
+	gcs.client = client
+
+}
+
+func (gcs *GCS) createGCSBucket(projectID string) {
+	// Setup client bucket to work from
+	gcs.bucket = gcs.client.Bucket(gcs.bucketName)
+
+	buckets := gcs.client.Buckets(gcs.ctx, projectID)
+	for {
+		attrs, err := buckets.Next()
+		// Assuming that if end Iterator then not found and need to create
+		if err == iterator.Done {
+			// Create bucket if it doesn't exist - https://cloud.google.com/storage/docs/reference/libraries
+			if err := gcs.bucket.Create(gcs.ctx, projectID, &storage.BucketAttrs {
+				Location:     "US",
+			}); err != nil {
+				// TODO - add random number to append to bucket name to resolve
+				log.Fatalf("Failed to create bucket:", err)
+			}
+			fmt.Printf("Bucket %v created.\n", gcs.bucketName)
+			return
+		}
+		if err != nil {
+			log.Fatalf("Issues setting up Bucket(%q).Objects(): %v. Double check project id.", attrs.Name, err)
+		}
+		if attrs.Name == gcs.bucketName{
+			//getLatestFile() // TODO set this up to check and compare what is in the bucket vs what isn't
+			fmt.Printf("Bucket %v exists.\n", gcs.bucketName)
+			return
+		}
+	}
+}
+
+func (gcs *GCS) storeGCS(fname string, url string) {
+	response, _ := getHTTPResponse(url)
+	if response.StatusCode == http.StatusOK {
+		obj := gcs.bucket.Object(fname)
+
+		// w implements io.Writer.
+		w := obj.NewWriter(gcs.ctx)
+
+		// Copy file into GCS
+		_, err := io.Copy(w, response.Body)
+		if err != nil {
+			log.Fatalf("Failed to copy doc to bucket:", err)
+		}
+		response.Body.Close()
+
+		// Close, just like writing a file.
+		if err := w.Close(); err != nil {
+			log.Fatalf("Failed to close:", err)
+		}
+	}
+}
+
+func getHTTPResponse(url string) (*http.Response, error){
+	response, err := http.Get(url)
+	if err != nil {
+		fmt.Println("****** HTTP ERROR *********", err)
+	}
+	// Defer response.Body.Close() // TODO figure out if this is usable in a func because its closing when passed
+	return response, err
+}
+
+
+func getMalingListData(url string, gcs GCS) {
```

Comment:
Should this be `getMailingListData`?

Reply:
Are you asking to change so the func name to match the file name?

Reply:
No -- there's a misspelling. You're missing an "i" in "Mailing".

Reply:
Couldn't see that - thanks for catching